waku/v2/protocol/waku_relay.nim

```diff
@@ -23,79 +27,132 @@ logScope:
       # do not need to propogate CancelledError.
-      trace "Unexpected cancellation in relay handler", conn
-    except CatchableError as exc:
-      trace "WakuRelay handler leaks an error", exc = exc.msg, conn
+      error "Unexpected cancellation in relay handler", conn=conn, error=getCurrentExceptionMsg()
+    except CatchableError:
+      error "WakuRelay handler leaks an error", conn=conn, error=getCurrentExceptionMsg()
 
   # XXX: Handler hijack GossipSub here?
   w.handler = handler
   w.codec = WakuRelayCodec
 
-method initPubSub*(w: WakuRelay) {.raises: [Defect, InitializationError].} =
-  debug "initWakuRelay"
+method initPubSub(w: WakuRelay) {.raises: [InitializationError].} =
+  ## NOTE: This method overrides GossipSub initPubSub method; it called by the 
+  ##  parent protocol, PubSub.
+  debug "init waku relay"
 
-  # after discussions with @sinkingsugar, this is essentially what is needed for
+  # After discussions with @sinkingsugar: This is essentially what is needed for
   # the libp2p `StrictNoSign` policy
   w.anonymize = true
   w.verifySignature = false
   w.sign = false
 
   procCall GossipSub(w).initPubSub()
 
-  w.init()
+  w.initProtocolHandler()
+
+
+proc new*(T: type WakuRelay, 
+          peerManager: PeerManager, 
+          defaultPubsubTopics: seq[PubsubTopic] = @[],
+          triggerSelf: bool = true): WakuRelayResult[T] =
+  
+  proc msgIdProvider(msg: messages.Message): Result[MessageID, ValidationResult] =
+    let hash = MultiHash.digest("sha2-256", msg.data)
+    if hash.isErr():
+      ok(($msg.data.hash).toBytes())
+    else:
+      ok(hash.value.data.buffer)
+
+  var wr: WakuRelay
+  try:
+    wr = WakuRelay.init(
+      switch = peerManager.switch,
+      msgIdProvider = msgIdProvider,
+      triggerSelf = triggerSelf,
+      sign = false,
+      verifySignature = false,
+      maxMessageSize = MaxWakuMessageSize
+    )
+  except InitializationError:
+    return err("initialization error: " & getCurrentExceptionMsg())
+
+  wr.peerManager = peerManager
+  wr.defaultPubsubTopics = defaultPubsubTopics
+
+  ok(wr)
+
+
+method start*(w: WakuRelay) {.async.} =
+  debug "start"
+  await procCall GossipSub(w).start()
+
+method stop*(w: WakuRelay) {.async.} =
+  debug "stop"
+  await procCall GossipSub(w).stop()
 
-method subscribe*(w: WakuRelay,
-                  pubSubTopic: PubsubTopic,
-                  handler: TopicHandler) =
-  debug "subscribe", pubSubTopic=pubSubTopic
 
-  procCall GossipSub(w).subscribe(pubSubTopic, handler)
+method subscribe*(w: WakuRelay, pubsubTopic: PubsubTopic, handler: SubsciptionHandler|PubsubRawHandler) =
+  debug "subscribe", pubsubTopic=pubsubTopic
 
-method publish*(w: WakuRelay,
-                pubSubTopic: PubsubTopic,
-                message: seq[byte]
-               ): Future[int] {.async.} =
-  trace "publish", pubSubTopic=pubSubTopic, message=message
+  var subsHandler: PubsubRawHandler 
```

Comment:
just a question. why `PubsubRawHandler`. don't we want to only deal with `WakuMessage`s?

Reply:
If I change to a unique `WakuMessage` handler, I would have to change half of the codebase (e.g., JSON-RPC, RLN relay, more WakuNode, etc.)

That's why we are supporting both options (for now).

Reply:
i see thanks. any opinion on wether we should support just wakumessage in the future?

Reply:
> any opinion on wether we should support just wakumessage in the future?

Not sure 100%; hence the dual handler support.